howl/client/howl_client.py

```diff
@@ -0,0 +1,140 @@
+
+    @staticmethod
+    def _get_device(device: int):
+        if device == -1:
+            return torch.device('cpu')
+        else:
+            return torch.device('cuda:{}'.format(device))
+
+    def _on_audio(self, in_data, frame_count, time_info, status):
+        data_ok = (in_data, pyaudio.paContinue)
+        self.last_data = in_data
+        self._audio_buf.append(in_data)
+        if len(self._audio_buf) != self._audio_buf_len:
+            return data_ok
+
+        audio_data = b''.join(self._audio_buf)
+        self._audio_buf = self._audio_buf[2:]
+        arr = self._normalize_audio(audio_data)
+        inp = torch.from_numpy(arr).float().to(self.device)
+
+        # Inference from input sequence
+        if self.engine.infer(inp):
+            phrase = ' '.join(self.ctx.vocab[x]
+                              for x in self.engine.sequence).title()
+            logging.info(f'{phrase} detected')
+            # Execute user-provided listener callbacks
+            for lis in self.listeners:
+                lis(self.engine.sequence)
+
+        return data_ok
+
+    def _normalize_audio(self, audio_data):
+        return np.frombuffer(audio_data, dtype=np.int16).astype(np.float) / self._audio_float_size
+
+    def start(self):
+        """Start the audio stream for inference"""
+        if self.engine is None:
+            raise AttributeError(
+                'Please provide an InferenceEngine or initialize using from_pretrained.'
+            )
+        if self.ctx is None:
+            raise AttributeError(
+                'Please provide an InferenceContext or initialize using from_pretrained.'
+            )
+
+        chosen_idx = 0
+        self._audio = pyaudio.PyAudio()
+        for idx in range(self._audio.get_device_count()):
+            info = self._audio.get_device_info_by_index(idx)
+            if info['name'] == 'pulse':
+                chosen_idx = idx
+                break
+
+        stream = self._audio.open(format=pyaudio.paInt16,
+                                  channels=1,
+                                  rate=16000,
+                                  input=True,
+                                  input_device_index=chosen_idx,
+                                  frames_per_buffer=self.chunk_size,
+                                  stream_callback=self._on_audio)
+        self.stream = stream
+        logging.info("Starting Howl inference client...")
+        stream.start_stream()
+        return self
+
+    def join(self):
+        """Block while the audio inference stream is active"""
+        while self.stream.is_active():
+            time.sleep(0.1)
+
+    def from_pretrained(self, name: str, force_reload: bool = False):
+        """Load a pretrained model using the provided name"""
+        engine, ctx = torch.hub.load(
+            'castorini/howl:howl-pip', name, force_reload=force_reload, reload_models=force_reload)
+        self.engine = engine.to(self.device)
+        self.ctx = ctx
```

Comment:
I know we had two parameters incorrectly setup previously.
`INFERENCE_THRESHOLD=0 NUM_MELS=40`
we should be loading these directly from these pre-trained models